### PR TITLE
Split squared norm equality check into a separate FLP

### PR DESCRIPTION
**Pull request by albertpl**
This implements #53.

### CODE REVIEW

**Comment by albertpl** [comment]
I suppose we need thorough security review on this approach. 
The intuition is that `eval_norm_check` doesn't depend on joint randomness. Although we still use joint randomness to combine results of all sub-circuits, the degree of the joint randomness doesn't change. So this won't make it easy to guess the joint randomness?

**Review by cjpatton — COMMENTED** [review]
This like it adds a flag to make the squared norm equality and range checks optional? That's not quite what we need.

The circuit checks for five properties, as shown here:
https://github.com/junyechen1996/draft-chen-cfrg-vdaf-pine/blob/f06d5344b0b9879d678e9b5ea01878b9a899dd9c/poc/flp_pine.py#L230-L234

We always need to check all five of these properties: it's not acceptable to skip the squared norm equality or range check under any circumstances.

The idea behind #53 is that that we can split out the squared norm equality check into its own circuit. We would then generate FLPs for two circuits: one for the squared norm equality check and another for all the other checks. This saves us computation because the squared norm equality check can be done without joint randomness, so we only need to generate one proof for it.

**Comment by junyechen1996** [comment]
> This like it adds a flag to make the squared norm equality and range checks optional? That's not quite what we need.
> 
> The circuit checks for five properties, as shown here:
> 
> https://github.com/junyechen1996/draft-chen-cfrg-vdaf-pine/blob/f06d5344b0b9879d678e9b5ea01878b9a899dd9c/poc/flp_pine.py#L230-L234
> 
> We always need to check all five of these properties: it's not acceptable to skip the squared norm equality or range check under any circumstances.
> 
> The idea behind #53 is that that we can split out the squared norm equality check into its own circuit. We would then generate FLPs for two circuits: one for the squared norm equality check and another for all the other checks. This saves us computation because the squared norm equality check can be done without joint randomness, so we only need to generate one proof for it.

Agreed. The current interface for `PineValid` makes it seem like norm equality check is optional. I think we should have one `Valid` circuit that just does the squared norm equality check which doesn't need joint randomness, something like `PineValidSquaredNormEqualityCheck`, and another `Valid` circuit that computes all other checks that require joint randomness, something like `PineValidOtherChecks`..?

**Comment by cjpatton** [comment]
Confirmed! Note you'll need to rebase this PR.

**Comment by albertpl** [comment]
Still WIP.  
I will be out of office for the rest of the week. So target to get this ready for review early next week.

**Comment by cjpatton** [comment]
Thanks @albertpl, this is heading in the right direction :)

**Comment by albertpl** [comment]
Rewrite this PR and send out for feedback. Pls take another look.

File: poc/Makefile [hunk 1]
```diff
@@ -1,2 +1,29 @@
-all:
+.PHONY: install lint tests all
+
+lint/pyflakes:
```
Comment (albertpl) [inline poc/Makefile:3]: How do you guys run these linting tools?

File: poc/Makefile [hunk 2]
```diff
@@ -1,2 +1,29 @@
...
 	sage -python -m unittest
+
+tests/mypy:
+	sage -python -m mypy *.py tests/*.py
+
+tests: tests/unittest tests/mypy
+
+install:
+	sage -python -m pip install pyflakes autopep8 isort pylint
+	sage -pip install git+https://github.com/cfrg/draft-irtf-cfrg-vdaf@fefc5a2f563e70e73a74d3c57b7145e0dd1c2e3f#subdirectory=poc
```
Comment (albertpl) [inline poc/Makefile:27]: Based on [this](https://github.com/cfrg/draft-irtf-cfrg-vdaf/pull/382) PR.

File: poc/flp_pine.py [hunk 3]
```diff
@@ -18,11 +19,18 @@
...
-            F,
-        ]):
+    Valid[
+        list[float],  # Measurement
+        list[float],  # AggResult
+        F,
+    ],
+    metaclass=ABCMeta
+):
+    """
```
Comment (albertpl) [inline poc/flp_pine.py:29]: Alternatively we can factor the common codes into some utility class and use "has-a" relationship.

File: poc/vdaf_pine.py [hunk 4]
```diff
@@ -105,26 +108,39 @@ def __init__(self,
                  l2_norm_bound: int,
                  num_frac_bits: int,
                  dimension: int,
-                 chunk_length: int,
+                 chunk_length_squared_norm_equality: int,
```
Comment (albertpl) [inline poc/vdaf_pine.py:111]: The optimal `chunk_length` can be quite different between `PineValidSquaredNormEqualityCheck` and `PineValidOtherChecks`. It makes senses to introduce a new parameter?

File: poc/Makefile (hunk 1 above)
Comment (cjpatton) [reply poc/Makefile:3]: I think most of them are available through `brew`.

File: poc/Makefile (hunk 2 above)
Comment (cjpatton) [reply poc/Makefile:27]: I'm worried about this Makefile diverging from the CI definitions. We should either:
1. Have CI call this Makefile
2. Have no Makefile at all.

I'd prefer the second one. If you want the Makefile for convenience, then you could include this in your local directory and add it to .gitignore so you don't accidentally check it in.

File: poc/flp_pine.py (hunk 3 above)
Comment (cjpatton) [reply poc/flp_pine.py:29]: What's your goal here?

File: poc/Makefile (hunk 1 above)
Comment (albertpl) [reply poc/Makefile:3]: Not installation part. So we just manually invoke those commands?

File: poc/Makefile (hunk 2 above)
Comment (albertpl) [reply poc/Makefile:27]: So we ask users to figure out how to run these steps by reading the CI workflow? And possibly reading man page for each command? e.g. 

In `workflow/lint-python.yml`, it runs the command with `--diff --exit-code`
`autopep8 --diff --exit-code *.py tests/*.py`

but we can use `--in-place` to automatically update
`sage -python -m autopep8 -i *.py tests/*.py`


In any case, this is unrelated changes. I will revert them and start another thread.

File: poc/flp_pine.py (hunk 3 above)
Comment (albertpl) [reply poc/flp_pine.py:29]: Just want to mention the alternative approach I have considered.

Comment (junyechen1996) [reply poc/flp_pine.py:29]: I think we can stick with the approach in the base VDAF poc.

**Review by junyechen1996 — APPROVED** [review]

File: poc/flp_pine.py [hunk 5]
```diff
@@ -503,17 +458,146 @@ def decode_float_from_field(self, field_elem: Field) -> float:
...
+        result = self.eval_norm_equality_check(
+            encoded_gradient, sq_norm_v_bits)
+        return [result]
+
+    def test_vec_set_type_param(self, test_vec):
+        test_vec["chunk_length_squared_norm_equality"] = self.chunk_length
+        return ["chunk_length_squared_norm_equality"]
+
+
+class PineValidOtherChecks(
```
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/flp_pine.py:507]: So you suggest we rename the base class, perhaps from `PineValid` -> `PineValidBase` or something else?

File: poc/flp_pine.py [hunk 6]
```diff
@@ -246,32 +202,29 @@ def eval_bit_checks(self,
...
-                        shares_inv: F) -> tuple[F, F]:
-        """
-        Check that (1) the reported squared was computed correctly and (2) the
-        squared norm is in range. The result is only valid if the bit checks
-        and the wraparound checks were successful.
-        """
+    def eval_norm_equality_check(self,
+                                 encoded_gradient: list[F],
+                                 sq_norm_v_bits: list[F]) -> F:
+        """Check that the computed squared L2-norm result matches the value claimed by the Client. """
```
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/flp_pine.py:208]: Is this enforced in any linting rule?
> Wrap lines at 80 characters

[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/flp_pine.py:208]: It doesn't appear to be at the moment.

File: poc/Makefile (hunk 1 above)
Comment (cjpatton) [reply poc/Makefile]: Yes.

File: poc/Makefile (hunk 2 above)
Comment (cjpatton) [reply poc/Makefile]: At the moment the "users" are just us :) I think for now we shouldn't try over do it on tooling, since this is likely to change quickly while we develop. Less is more?

File: poc/flp_pine.py [hunk 7]
```diff
@@ -36,10 +44,7 @@ def __init__(self,
                  num_wr_checks: int = NUM_WR_CHECKS,
                  num_wr_successes: int = NUM_WR_SUCCESSES):
         """
-        Instantiate the `PineValid` circuit for gradients with `dimension`
-        elements. Each element will be truncated to `num_frac_bits` binary
-        fractional bits, and the L2-norm bound of each gradient is bounded
-        by `l2_norm_bound`.
+        Callers should explicitly instantiate its subclasses, instead of calling this method.
```
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/flp_pine.py:47]: I split & moved these details to its subclasses, s.t. we don't have to duplicate these documents. What do you think?

File: poc/flp_pine.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/flp_pine.py:507]: `PineValidBase` sounds reasonable.

File: poc/tests/test_flp.py [hunk 8]
```diff
@@ -254,27 +256,27 @@ def test_flp(self):
...
-            test_flp_bbcggi19(flp, [(meas, True)])
+                # Test PINE FLP with verification.
+                xof = XofTurboShake128(gen_rand(16), b"", b"")
+                encoded_gradient_and_norm = \
+                    flp.valid.encode_gradient_and_norm(measurement)
+                (wr_check_bits, wr_check_results) = \
+                    pine_valid.encode_wr_checks(encoded_gradient_and_norm[:dimension],
+                                                xof)
+                meas = encoded_gradient_and_norm + wr_check_bits + wr_check_results
+                test_flp_bbcggi19(flp, [(meas, True)])
```
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/tests/test_flp.py:279]: Encoding requires an instant of `flp`, which is created inside the loop?
So we should test with one `flp` to encode and then create another `flp` to test the rest. A little bit weird?

File: poc/flp_pine.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/flp_pine.py]: OK. I add `--max-line` to my local config.

**Comment by albertpl** [comment]
> It seems like the pattern you want to follow is: PineValid implements the functionality shared by both circuits, then sub-classes implement the circuit-specific functionality. This pattern might make sense, but as it is, I think base class PineValid is doing too much:
Not all parameters set by the constructor are relevant to both circuits
The base class includes the sub-circuits that are used by PineValidOtherChecks but not by PineValidSquaredNormEqualityCheck
The base class includes encoding routines that are not relevant to both
I'd like to see a cleaner separation of the two circuits.


* I relocated all the evaluation sub-circuits.
* but the encoding routines are shared by both circuits such that both can be used individually.
* most of parameters are shared by both too. One reason is that we initialize them together, with dependencies on each other. Two initialization of these parameters would be a little bit tricky? Lots of the parameters are used by the shared routines as well, e.g. `unpack_encoded_measurement`.

File: poc/flp_pine.py (hunk 3 above)
Comment (albertpl) [reply poc/flp_pine.py:29]: * I relocated all the evaluation sub-circuits.
* But the encoding routines are shared by both circuits such that both can be used individually.
* Also Most of parameters are shared by both too. One reason is that we initialize them together, with dependencies on each other. Two initialization of these parameters would be a little bit tricky? Lots of the parameters are used by the shared routines as well, e.g. unpack_encoded_measurement.

File: poc/flp_pine.py [hunk 9]
```diff
@@ -503,17 +362,257 @@ def decode_float_from_field(self, field_elem: Field) -> float:
...
+
+    def test_vec_set_type_param(self, test_vec):
+        test_vec["chunk_length_squared_norm_equality"] = self.chunk_length
+        return ["chunk_length_squared_norm_equality"]
+
+
+class PineValid(
+    PineValidBase[F]
+):
+    """Validity circuit for all other checks in PINE, including
```
Comment (cjpatton) [inline poc/flp_pine.py:431]:
```suggestion
        PineValidBase[F]
    ):
    """
    Validity circuit for all other checks in PINE, including
```

Comment (cjpatton) [inline poc/flp_pine.py:431]:
```suggestion
    PineValidBase[F]
):
    """Validity circuit for all checks for PINE except the squared norm quality check. This includes
```

File: poc/tests/test_flp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/tests/test_flp.py:282]: Weird may be preferable. We know that the encoding step is expensive, so we might save a lot on unit tests if we only encode once.

CI takes something like 10 seconds, which is quite slow for unittests. I'm not sure though if this change is the reason.

File: poc/vdaf_pine.py [hunk 10]
```diff
@@ -105,12 +108,23 @@ def __init__(self,
...
+            field,
+            l2_norm_bound,
+            num_frac_bits,
+            dimension,
+            chunk_length_squared_norm_equality,
+            alpha,
+            num_wr_checks,
+            num_wr_successes,
+        )
         self.valid = PineValid(
```
Comment (cjpatton) [inline poc/vdaf_pine.py:128]: Idea: it might be nice to construct these FLPs once. Could we do something like this:
```python
(self.valid_squared_norm_equality, self.valid) = pine_flp.construct_circuits(field, l2_Norm_bound, ...)
```

File: poc/vdaf_pine.py [hunk 11]
```diff
@@ -121,6 +135,8 @@ def __init__(self,
             num_wr_checks,
             num_wr_successes,
         )
+        self.flp_squared_norm_equality = FlpBBCGGI19(
+            self.valid_squared_norm_equality)
```
Comment (cjpatton) [inline poc/vdaf_pine.py:139]: Do we actually need to store `valid_squared_norm_equality` in `self`?

File: poc/vdaf_pine.py [hunk 12]
```diff
@@ -471,6 +512,7 @@ def helper_proofs_share(self,
             k_share,
             self.domain_separation_tag(USAGE_PROOF_SHARE),
             byte(self.PROOFS) + byte(agg_id),
+            self.flp_squared_norm_equality.PROOF_LEN +
             self.flp.PROOF_LEN * self.PROOFS
```
Comment (cjpatton) [inline poc/vdaf_pine.py:516]: This is visually hard to parse: the two lines look like different input parameters. Here and below.
```suggestion
            self.flp_squared_norm_equality.PROOF_LEN +
                self.flp.PROOF_LEN * self.PROOFS
```

**Review by cjpatton — COMMENTED** [review]
Looks good overall! Just some nit-picky stuff

File: poc/tests/test_flp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/tests/test_flp.py:282]: How much we can save here? `dimension=4`. I can reduce num_wr_check if that really matter

File: poc/flp_pine.py [hunk 13]
```diff
@@ -377,7 +233,8 @@ def encode_gradient_and_norm(self, measurement: list[float]) -> list[F]:
                 self.num_bits_for_sq_norm,
             )
 
-    def run_wr_checks(self, encoded_gradient: list[F], wr_joint_rand_xof: Xof) -> list[F]:
+    def run_wr_checks(
```
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/flp_pine.py:236]: We also need `unpack_encoded_measurement` which depends on lots of parameters. Unless we want to duplicate this in both `PineSquaredNormEqualityCheck` and `PineValid`

File: poc/vdaf_pine.py (hunk 11 above)
Comment (albertpl) [reply poc/vdaf_pine.py]: removed

File: poc/vdaf_pine.py (hunk 10 above)
Comment (albertpl) [reply poc/vdaf_pine.py]: done

File: poc/flp_pine.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/flp_pine.py]: I move all encoding functions to `PineValid`.

File: poc/tests/test_flp.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/tests/test_flp.py]: nvm. done.

File: poc/flp_pine.py [hunk 14]
```diff
@@ -195,6 +174,170 @@ def eval(self,
...
+        return ["chunk_length_norm_equality"]
+
+
+class PineValid(
+    PineValidBase[F]
+):
+    """
+    Validity circuit for all checks for PINE except the squared norm quality 
+    check. The result is only valid if the circtuit 
+    `PineValidSquaredNormEqualityCheck` is also valid.
```
Comment (cjpatton) [inline poc/flp_pine.py:293]: spelling and trailing whitespace
```suggestion
    Validity circuit for all checks for PINE except the squared norm quality
    check. The result is only valid if the circuit
    `PineValidSquaredNormEqualityCheck` is also valid.
```

File: poc/flp_pine.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/flp_pine.py]: Awesome, I think that makes more sense.

File: poc/tests/test_flp.py [hunk 15]
```diff
@@ -265,16 +266,19 @@ def test_flp(self):
...
             encoded_gradient_and_norm = \
                 flp.valid.encode_gradient_and_norm(measurement)
             (wr_check_bits, wr_check_results) = \
                 pine_valid.encode_wr_checks(encoded_gradient_and_norm[:dimension],
                                             xof)
             meas = encoded_gradient_and_norm + wr_check_bits + wr_check_results
-            test_flp_bbcggi19(flp, [(meas, True)])
+
+            for Valid in [PineValid, PineValidSquaredNormEqualityCheck]:
+                test_flp = FlpBBCGGI19(Valid(field, *args))
```
Comment (cjpatton) [inline poc/tests/test_flp.py:278]: nit: call `construct_circuits()` and iterate over the outputs:
```python
circuits = construct_circuits(BLAH)
for v in circuits:
    BLAH
```

File: poc/tests/test_vdaf.py [hunk 16]
```diff
@@ -40,6 +40,7 @@ def test_result_share_length(self):
                       num_frac_bits=4,
                       dimension=4,
                       chunk_length=150,
+                      chunk_length_norm_equality=2,
```
Comment (cjpatton) [inline poc/tests/test_vdaf.py:43]: Make sure the code is consistent about "squared norm equality" versus "norm equality". Either is fine with me, but consistency is important for readability.

File: poc/flp_pine.py [hunk 17]
```diff
@@ -549,3 +635,35 @@ def chunk_count(chunk_length, length):
...
+    ), PineValid(
+        field,
+        l2_norm_bound,
+        num_frac_bits,
+        dimension,
+        chunk_length,
+        alpha,
+        num_wr_checks,
+        num_wr_successes,
+    )
```
Comment (cjpatton) [inline poc/flp_pine.py:669]: Idea: We could move the code from `PineValiBase.__init__` here. Then all each circuit's constructor needs to do is copy the parameters it needs and construct the ones that are specific to it.

File: poc/tests/test_vdaf.py (hunk 16 above)
Comment (albertpl) [reply poc/tests/test_vdaf.py:43]: I stick with "norm equality".

Comment (albertpl) [reply poc/tests/test_vdaf.py:43]: Rename `PineValidSquaredNormEqualityCheck` to `PineValidNormEquality`. 
The comments may still refer to "squared norm check equality", let me know you don't like this as well.

Comment (cjpatton) [reply poc/tests/test_vdaf.py:43]: Comments are fine, thanks

File: poc/flp_pine.py (hunk 17 above)
Comment (albertpl) [reply poc/flp_pine.py:669]: Both critics share lots of parameters. Unless we want these two circuits to have different `MEAS_LEN`,  `PineValidNormEquality` has to carry `num_wr_checks`. So goes to other parameters.

What parameters do you want to split out of base class?


The base class should initializer the shared parameters. Having some external method to do this work is not a good practice.

And each constructor of the subclass already does "construct the ones that are specific to it." as of now.

File: poc/flp_pine.py [hunk 18]
```diff
@@ -503,17 +362,257 @@ def decode_float_from_field(self, field_elem: Field) -> float:
         decoded_float = decoded / (2 ** self.num_frac_bits)
         return decoded_float
 
+
+class PineValidSquaredNormEqualityCheck(
+    PineValidBase[F]
+):
```
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/flp_pine.py]: I tried but the lint step automatically revert this change. Let me bypass the lint step.

File: poc/flp_pine.py [hunk 19]
```diff
@@ -195,6 +174,168 @@ def eval(self,
...
+
+
+class PineValidNormEquality(
+        PineValidBase[F]
+    ):
+    """
+    PINE validation circuit that computes the squared L2-norm result
+    and checks that it matches the value claimed by the Client.
+        * This circuit does not use joint randomness.
+        * The result is only valid if the other circuit
```
Comment (cjpatton) [inline poc/flp_pine.py:232]: trailing whitespace
```suggestion
        * The result is only valid if the other circuit
```

**Review by cjpatton — APPROVED** [review]
LGTM modulo one byte of trailing whitespace

File: poc/tests/test_flp.py (hunk 15 above)
Comment (albertpl) [reply poc/tests/test_flp.py]: done

File: poc/flp_pine.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (albertpl) [reply poc/flp_pine.py]: This does fail [autopep8 check](https://github.com/junyechen1996/draft-chen-cfrg-vdaf-pine/actions/runs/10323919100/job/28582293557?pr=92) the lint check. So I have to revert.

N.B. this is automatically suggested by `autopep8 --diff --exit-code *.py tests/*.py`
And

[the remark this answers is not in the document]
Comment (cjpatton) [reply poc/flp_pine.py]: Oh no, sorry about the conflicting feedback!
